daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ComparisonSBuiltinTest.scala

```diff
@@ -0,0 +1,645 @@
+              e"Mod:Tuple @Int64 @Int64 {fst = 1, snd = 0}",
+              e"Mod:Tuple @Int64 @Int64 {fst = 1, snd = 1}"
+            ),
+          t"Mod:Either Text Int64" -> List(
+            e"""Mod:Either:Left @Text @Int64 "a" """,
+            e"""Mod:Either:Left @Text @Int64 "b" """,
+            e"""Mod:Either:Right @Text @Int64 -1 """,
+            e"""Mod:Either:Right @Text @Int64 1 """,
+          ),
+          t"<fst: Int64, snd: Int64>" ->
+            List(
+              e"<fst = 0, snd = 0>",
+              e"<fst = 0, snd = 1>",
+              e"<fst = 1, snd = 0>",
+              e"<fst = 1, snd = 1>"
+            ),
+          t"Optional Text" ->
+            List(
+              e""" None @Text """,
+              e""" Some @Text "A" """,
+              e""" Some @Text "AA" """,
+            ),
+          t"List Int64" ->
+            List(
+              e"Nil @Int64",
+              e"Cons @Int64 [0] (Nil @Int64)",
+              e"Cons @Int64 [0,0] (Nil @Int64)",
+              e"Cons @Int64 [0,0,1] (Nil @Int64)",
+              e"Cons @Int64 [0,1,1] (Nil @Int64)",
+              e"Cons @Int64 [1] (Nil @Int64)",
+              e"Cons @Int64 [1,0] (Nil @Int64)",
+              e"Cons @Int64 [1,1] (Nil @Int64)",
+            ),
+          t"TextMap Int64" ->
+            List(
+              e"TEXTMAP_EMPTY @Int64",
+              e"""TEXTMAP_INSERT @Int64 "a" 1 (TEXTMAP_EMPTY @Int64)""",
+              e"""TEXTMAP_INSERT @Int64 "a" 1 (TEXTMAP_INSERT @Int64"b" 1 (TEXTMAP_EMPTY @Int64))""",
+              e"""TEXTMAP_INSERT @Int64 "a" 1 (TEXTMAP_INSERT @Int64"b" 2 (TEXTMAP_EMPTY @Int64))""",
+              e"""TEXTMAP_INSERT @Int64 "a" 2 (TEXTMAP_EMPTY @Int64)""",
+              e"""TEXTMAP_INSERT @Int64 "a" 2 (TEXTMAP_INSERT @Int64 "b" 1 (TEXTMAP_EMPTY @Int64))""",
+              e"""TEXTMAP_INSERT @Int64 "a" 2 (TEXTMAP_INSERT @Int64 "b" 2 (TEXTMAP_EMPTY @Int64))""",
+              e"""TEXTMAP_INSERT @Int64 "a" 2 (TEXTMAP_INSERT @Int64 "b" 2 (TEXTMAP_INSERT @Int64 "c" 3 (TEXTMAP_EMPTY @Int64)))""",
+            ),
+          t"GenMap Text Int64" ->
+            List(
+              e"GENMAP_EMPTY @Text @Int64",
+              e"""GENMAP_INSERT @Text @Int64 "a" 1 (GENMAP_EMPTY @Text @Int64)""",
+              e"""GENMAP_INSERT @Text @Int64 "a" 1 (GENMAP_INSERT @Text @Int64"b" 1 (GENMAP_EMPTY @Text @Int64))""",
+              e"""GENMAP_INSERT @Text @Int64 "a" 1 (GENMAP_INSERT @Text @Int64"b" 2 (GENMAP_EMPTY @Text @Int64))""",
+              e"""GENMAP_INSERT @Text @Int64 "a" 2 (GENMAP_EMPTY @Text @Int64)""",
+              e"""GENMAP_INSERT @Text @Int64 "a" 2 (GENMAP_INSERT @Text @Int64 "b" 1 (GENMAP_EMPTY @Text @Int64))""",
+              e"""GENMAP_INSERT @Text @Int64 "a" 2 (GENMAP_INSERT @Text @Int64 "b" 2 (GENMAP_EMPTY @Text @Int64))""",
+              e"""GENMAP_INSERT @Text @Int64 "a" 2 (GENMAP_INSERT @Text @Int64 "b" 2 (GENMAP_INSERT @Text @Int64 "c" 3 (GENMAP_EMPTY @Text @Int64)))""",
```

Comment:
probably we could try to factorize that.
Note, we still have to dynamically preserve the type since the conversion to the external world does not know the type (cf. `SValue.toValue`).